org.eclipse.xtext.xbase/src/org/eclipse/xtext/xbase/compiler/AbstractXbaseCompiler.java

```diff
@@ -345,7 +346,7 @@ protected boolean hasJvmConstructorCall(XExpression obj) {
 	protected boolean needsSneakyThrow(XExpression obj, Collection<JvmTypeReference> declaredExceptions) {
 		IResolvedTypes resolvedTypes = getResolvedTypes(obj);
 		List<LightweightTypeReference> thrownExceptions = resolvedTypes.getThrownExceptions(obj);
```

Comment:
The thrown exceptions should already include the exception declared on the AutoClosable.

Reply:
Indeed.
However when the resource is defined with a lambda expression, it is assumed to be an AutoClosable, but the corresponding exceptions are not listed in resolvedTypes.getThrownExceptions(obj). For that case I implemented the additional method.

Should I refine the method to that specific case, that the right hand side of the resource definition is a lambda expression?


example:
```
try (var r = [System.out.println("Closing")]) {} 
```
would generate the Java Code:
```
try (AutoCloseable r = ((AutoCloseable) () -> {
   System.out.println("Closing");
})) {}
```
and is marked as error, that AutoClosable throws an unhandled excepetion.
With the additional method is would result in
```
try {
   try (AutoCloseable r = ((AutoCloseable) () -> {
      System.out.println("Closing");
   })) {
   }
} catch (Throwable _e) {
   throw Exceptions.sneakyThrow(_e);
}
```

Reply:
It should collect all the declared exceptions of the close methods of the resource vars

Reply:
I has some more thoughts on and experiments with this issue:

Indeed, all exceptions get collected, at least if `close()` is explicitly called within the try's expression block.
However (independent from if the rhs is a lambda expression or not), if the resource is considered a not further specified `AutoClosable` and if `close()` is not explicitly called, but implicitly by the try with resource statement, the java error is thrown, that the resource has unhandled Excpetions. See cases 1 and 2 below.

I'm not sure if I can find out how to convince the collector to collect an unhandled thrown excepetion in case the resource is an AutoClosable. The much easier but less clean solution is to stick to the initial attempt, using a method to check if the resource is an AutoClosable and define that it needs a SneakyThrow in that case.

Case 1:
The Xtend Code with `someClosable` as an AutoClosable does not call any methods with the expression block and thus is considered to not throw any exceptions. 
```
val reader1 = new StringReader("Text1")
val reader2 = new StringReader("Text2")
val array = #[reader1,reader2]
		
try (val AutoCloseable someClosable = array.get(0)){}
```

In the generated Java Code the error is marked: "Unhandled exception type Exception thrown by automatic close() invocation on someClosable"
```
final StringReader reader1 = new StringReader("Text1");
final StringReader reader2 = new StringReader("Text2");
final List<StringReader> array = Collections.<StringReader>unmodifiableList(CollectionLiterals.<StringReader>newArrayList(reader1, reader2));
try (final AutoCloseable someClosable = array.get(0)) {
}
```

Case 2:
If one calls `someClosable.close()` explicitly in the try expression, in the generated Java Code an Excpetion is catched:
```
try {
      final StringReader reader1 = new StringReader("Text1");
      final StringReader reader2 = new StringReader("Text2");
      final List<StringReader> array = Collections.<StringReader>unmodifiableList(CollectionLiterals.<StringReader>newArrayList(reader1, reader2));
      try (final AutoCloseable closableOhne = array.get(0)) {
        closableOhne.close();
      }
} catch (Throwable _e) {
      throw Exceptions.sneakyThrow(_e);
}
```

Case 3:
If someClosable is a StringReader no exception on close is thrown and thus a try withour a cacth does not have any errors. If one would call e.g. `read()` in the expression block, a catch clause would be generated.
Xtend Code:
```
try (val someClosable = new StringReader("Text")){}
```
JavaCode without error, even though not catch clause:
```
try (final StringReader someClosable = new StringReader("Text")) {
}
```

Case 4:
If someClosable is another implementation of AutoClosable, like FileInputStream, FileReader, etc, a catch clause is generated as well.
Xtend Code:
```
try (val someCloseable = new FileInputStream("test.txt")) {}
```
Java Code:
```
try {
    try (final FileInputStream someCloseable = new Function0<FileInputStream>() {
        public FileInputStream apply() {
          try {
            return new FileInputStream("test.txt");
          } catch (Throwable _e) {
            throw Exceptions.sneakyThrow(_e);
          }
        }
      }.apply()) {
      }
} catch (Throwable _e) {
      throw Exceptions.sneakyThrow(_e);
}
```

Thus I think it is legit to call for a sneakyThrow whenever a resource is an AutoClosable. It also does not hurt if it actually isn't needed (as it should be in case 2).

Reply:
One could add a query in `org.eclipse.xtext.xbase.typesystem.util.ThrownExceptionSwitch.caseXTryCatchFinallyExpression(XTryCatchFinallyExpression)`, where the catch clauses of a try statement are processed. In case of an AutoClosable (or XClosure) resource, an exception would be added to the list of thrown exceptions.
But I'm not convinced, if this attempt is cleaner than the simple one above.
And I'm not quite sure, how to add the exception.